src/main/java/com/pinterest/secor/parser/PartitionFinalizer.java

```diff
@@ -61,154 +60,123 @@ public PartitionFinalizer(SecorConfig config) throws Exception {
+        List<Message> committedMessages = new ArrayList<Message>(numPartitions);
         for (int partition = 0; partition < numPartitions; ++partition) {
             TopicPartition topicPartition = new TopicPartition(topic, partition);
-            long timestamp = getLastTimestampMillis(topicPartition);
-            if (timestamp > max_timestamp) {
-                max_timestamp = timestamp;
+            Message lastMessage = mKafkaClient.getLastMessage(topicPartition);
+            Message committedMessage = mKafkaClient.getCommittedMessage(topicPartition);
+            if (lastMessage == null || committedMessage == null) {
+                // This will happen if no messages have been posted to the given topic partition.
+                LOG.error("For topic {} partition {}, lastMessage: {}, commmitted: {}",
+                    topicPartition.getTopic(), topicPartition.getPartition(),
+                    lastMessage, committedMessage);
+                continue;
             }
+            lastMessages.add(lastMessage);
+            committedMessages.add(committedMessage);
         }
-        if (max_timestamp == Long.MIN_VALUE) {
-            return -1;
-        }
-        return max_timestamp;
+        return mMessageParser.getFinalizedUptoPartitions(lastMessages, committedMessages);
     }
 
-    private long getCommittedTimestampMillis(TopicPartition topicPartition) throws Exception {
-        Message message = mKafkaClient.getCommittedMessage(topicPartition);
-        if (message == null) {
-            LOG.error("No message found for topic {} partition {}", topicPartition.getTopic(), topicPartition.getPartition());
-            return -1;
-        }
-        return mMessageParser.extractTimestampMillis(message);
-    }
+    private void finalizePartitionsUpTo(String topic, String[] partitions) throws Exception {
+        final String s3Prefix = "s3n://" + mConfig.getS3Bucket() + "/" + mConfig.getS3Path();
 
-    private long getCommittedTimestampMillis(String topic) throws Exception {
-        final int numPartitions = mKafkaClient.getNumPartitions(topic);
-        long minTimestamp = Long.MAX_VALUE;
-        for (int partition = 0; partition < numPartitions; ++partition) {
-            TopicPartition topicPartition = new TopicPartition(topic, partition);
-            long timestamp = getCommittedTimestampMillis(topicPartition);
-            if (timestamp == -1) {
-                return -1;
-            } else {
-                if (timestamp < minTimestamp) {
-                    minTimestamp = timestamp;
-                }
-            }
-        }
-        if (minTimestamp == Long.MAX_VALUE) {
-            return -1;
-        }
-        return minTimestamp;
-    }
+        // Walk through each dimension of the partitions array
+        // First finalize the hourly partitions, and then move on to daily partitions
+        for (int dim = partitions.length; dim > 0; dim--) {
```

Comment:
If we do this, this is probably better done to have Partitioner to provide a Iterator class or simply returns a list of those partitions (bounded by look.back.periods), otherwise it would be hard for partitioner to keep the iterating state.

There are some differences between those partitions, e.g. dt=2015-07-01 don't need to do hive registration while the hourly partition needs to, I guess the finalizer needs to figure this out maybe by comparing the length of the those partition elements.